### PR TITLE
Allowed me page local operations

### PR DESCRIPTION
**Pull request by ZhangYihanNus**
able to get user avatar and nickname;
able to edit and save user's information locally;
problem of the malposed cursors in the input modules to be solved (no effect on operating);

### CODE REVIEW

File: client/app.js [hunk 1]
```diff
@@ -11,7 +11,37 @@ App({
     qcloud.login({
       success(res) {
         that.userInfo = res;
+        console.log(that.userInfo);
```
Comment (MaxPoon) [inline client/app.js]: Please remove all the loggings.

File: client/app.js [hunk 2]
```diff
@@ -11,7 +11,37 @@ App({
...
+
+        qcloud.request({
+          url: config.service.userUrl + '?OpenID=' + that.userInfo.openId,
+          data: {
```
Comment (MaxPoon) [inline client/app.js]: I don't think you need to send the data for a GET user request.

File: client/app.js [hunk 3]
```diff
@@ -11,7 +11,37 @@ App({
...
+            OpenID: that.userInfo.openId
+          },
+          method: 'GET', // OPTIONS, GET, HEAD, POST, PUT, DELETE, TRACE, CONNECT
+          
+          success: function (res) {
+            console.log(res);
```
Comment (MaxPoon) [inline client/app.js]: I think when it successfully retrieves the information, it should update `that.userInfo`.

File: client/app.js [hunk 4]
```diff
@@ -11,7 +11,37 @@ App({
...
+            console.log('complete wx.request')
+          }
+        });
+      },
+      fail (err) {
+        console.log('fail to login');
+        console.log(err)
       }
     });
+
```
Comment (MaxPoon) [inline client/app.js]: Please remove the extra empty lines.

File: client/app.wxss [hunk 5]
```diff
@@ -1,7 +1,7 @@
 /**app.wxss**/
 .container {
-  display: flex;
+  /* display: flex;
   flex-direction: column;
-  align-items: center;
+  align-items: center; */
```
Comment (MaxPoon) [inline client/app.wxss]: Should these changes be applied to the whole app or the me page only?

File: client/pages/me/me.js [hunk 6]
```diff
@@ -1,66 +1,71 @@
...
-  onShow: function () {
+  onLoad: function () {
+    console.log(app.userInfo)
+    this.setData({
+      userInfo: app.userInfo
+    })
 
+    if (this.data.userInfo == null) {
+        // if there is no userinfo yet
+        setTimeout(function () {}, 100)
```
Comment (MaxPoon) [inline client/pages/me/me.js]: Why do we do this?
I think a `setTimeout` function will return immediately. Only the callback function will be executed with a delay.

File: client/pages/me/me.js [hunk 7]
```diff
@@ -1,66 +1,71 @@
...
+    } 
   },
+ 
 
-  /**
-   * 生命周期函数--监听页面隐藏
-   */
-  onHide: function () {
-
+  chNameInput: function(event) {
```
Comment (MaxPoon) [inline client/pages/me/me.js]: Normally we want the variable names to be clearer. So `chineseNameInput`, `englishNameInput`, `chineseNameTemp `... look better to me.

File: client/pages/me/me.js [hunk 8]
```diff
@@ -1,66 +1,71 @@
...
+    this.setData({
+      chNameTemp: event.detail.value,
```
Comment (MaxPoon) [inline client/pages/me/me.js]: It's the input value of the field. How about calling it `chineseNameInputValue` instead?

File: client/app.js [hunk 9]
```diff
@@ -11,7 +11,37 @@ App({
...
       success(res) {
         that.userInfo = res;
+        console.log(that.userInfo);
+
+        qcloud.request({
+          url: config.service.userUrl + '?OpenID=' + that.userInfo.openId,
+          data: {
+            OpenID: that.userInfo.openId
+          },
+          method: 'GET', // OPTIONS, GET, HEAD, POST, PUT, DELETE, TRACE, CONNECT
```
Comment (MaxPoon) [inline client/app.js]: You can remove the comment.

File: client/pages/me/me.js [hunk 10]
```diff
@@ -1,66 +1,71 @@
-// pages/me/me.js
-Page({
+//index.js
+const app = getApp()
```
Comment (MaxPoon) [inline client/pages/me/me.js]: Missing semicolon.

File: client/pages/me/me.js [hunk 11]
```diff
@@ -1,66 +1,71 @@
...
 
-  /**
-   * 生命周期函数--监听页面显示
-   */
-  onShow: function () {
+  onLoad: function () {
+    console.log(app.userInfo)
+    this.setData({
+      userInfo: app.userInfo
+    })
```
Comment (MaxPoon) [inline client/pages/me/me.js]: Missing semicolon.

File: client/pages/me/me.wxml [hunk 12]
```diff
@@ -1,2 +1,62 @@
...
+        </view>
+        <!-- waiting for server link -->
+        <view wx:else><input placeholder-class='phText' placeholder='{{phone}}' bindinput='phoneInput'/></view>
+      </view>
+    </view>
+
+    <view class="info">   <!-- email -->
+      <text>E-mail</text>
+      <view class="infoInput">
+        <view wx:if="{{!email}}" >
```
Comment (MaxPoon) [inline client/pages/me/me.wxml]: Can we apply `wx:if` and `wx:else` directly to the input field, so that we don't need an extra view tag here?

File: project.config.json [hunk 13]
```diff
@@ -8,5 +8,6 @@
 	},
 	"appid": "wx13a399684987dfa0",
 	"projectname": "TPCC-YEC-Soccer-Team-Mini-Program",
+
```
Comment (MaxPoon) [inline project.config.json]: Revert the change here if it isn't needed.

File: client/pages/me/me.js [hunk 14]
```diff
@@ -1,66 +1,71 @@
...
+      saveBtnType: 'primary'
+    })
   },
 
-  /**
-   * 用户点击右上角分享
-   */
-  onShareAppMessage: function () {
+  infoSaveClick: function(event) {
+    this.setData({
```
Comment (MaxPoon) [inline client/pages/me/me.js]: I think we should update the user info on server via the `POST` endpoint.

Comment (MaxPoon) [reply client/pages/me/me.js]: And probably we should ignore the button click if `saveBtnType  !== 'default'`.

File: client/pages/me/me.js (hunk 10 above)
Comment (MaxPoon) [reply client/pages/me/me.js]: It's always safer to have a semicolon at the end of each statement. See the examples of errors causing by missing semicolons here: https://flaviocopes.com/javascript-automatic-semicolon-insertion/

**Review by MaxPoon — CHANGES_REQUESTED** [review]

File: client/pages/me/me.js [hunk 15]
```diff
@@ -1,66 +1,71 @@
-// pages/me/me.js
-Page({
+//index.js
```
Comment (MaxPoon) [inline client/pages/me/me.js]: Should be me.js

File: client/config.js [hunk 16]
```diff
@@ -3,7 +3,7 @@
  */
 
 // 此处主机域名修改成腾讯云解决方案分配的域名
-var host = 'https://aeoo5f7q.qcloud.la';
+var host = 'http://localhost:5757';
```
Comment (MaxPoon) [inline client/config.js]: We shouldn't change it to the local url.

File: client/pages/me/me.js [hunk 17]
```diff
@@ -1,66 +1,98 @@
...
   },
-
-  /**
-   * 生命周期函数--监听页面加载
-   */
-  onLoad: function (options) {
-
+  onLoad: function () {
+    this.setData({
+      personal: app.userInfo
```
Comment (MaxPoon) [inline client/pages/me/me.js:16]: LGTM for now. But eventually we need to handle the case that user info isn't available (network issue).

File: client/pages/me/me.wxml [hunk 18]
```diff
@@ -1,2 +1,54 @@
...
+    <text class="userinfo-nickname">{{userInfo.nickName}}</text>
+  </view>
+
+
+  <view class="body">   <!-- info -->
+    <view class="info">   <!-- Chinese name -->
+      <text>Chinese name</text>
+      <view class="infoInput">
+          <input wx:if="{{!personal.data.ChineseName}}" placeholder='Tap here to enter' bindinput='chineseNameInput'/>
+        <!-- waiting for server link -->
```
Comment (MaxPoon) [inline client/pages/me/me.wxml]: You can remove the comment now.

File: client/pages/me/me.js [hunk 19]
```diff
@@ -1,66 +1,98 @@
...
+    this.setData({
+      chineseName: this.data.chineseNameInputValue,
+      englishName: this.data.englishNameInputValue,
+      phone: this.data.phoneInputValue,
+      email: this.data.emailInputValue,
+      btnDisable: true
+    });
+    //update the latest info into server
+    qcloud.request({
+      url: 'http://localhost:5757',//config.service.userUrl + '?OpenID=' + that.userInfo.openId,
```
Comment (MaxPoon) [inline client/pages/me/me.js]: Please use `config.service.userUrl`

File: client/pages/me/me.js [hunk 20]
```diff
@@ -1,66 +1,98 @@
...
+      method: 'POST',
+      header: {
+        'content-type':'application/x-www-form-urlencoded'
+      },
+      data: {
+        //OpenID: this.personal.openId (using annotation to avoid error of undefined openID)
```
Comment (MaxPoon) [inline client/pages/me/me.js]: I think the OpenId field is needed.

File: client/pages/me/me.js [hunk 21]
```diff
@@ -1,66 +1,98 @@
...
+        ChineseName: this.data.chineseName,
+        EnglishName: this.data.englishName,
+        MobileNumber: this.data.phone,
+        EmailAddress: this.data.email
+      },
+      success: function (res) {
+        //able update userinfo into server
+        console.log('successed to update userinfo');
```
Comment (MaxPoon) [inline client/pages/me/me.js]: Please remove the loggings.

File: client/pages/me/me.wxml [hunk 22]
```diff
@@ -1,2 +1,54 @@
...
+        <input wx:else placeholder-class='phText' placeholder='{{chineseName}}' bindinput='chineseNameInput'/>
```
Comment (MaxPoon) [inline client/pages/me/me.wxml]: Instead of setting the placeholder, maybe we can set the value directly, so that we don't need two variables for each field (chineseName and chineseNameInputValue). It makes the code much simpler. What do you think?

File: client/pages/me/me.js [hunk 23]
```diff
@@ -1,66 +1,98 @@
...
   },
 
-  /**
-   * 生命周期函数--监听页面初次渲染完成
-   */
-  onReady: function () {
-
+  chineseNameInput: function (event) {
+    if (event.detail.value.length > 0) {
+      this.btnDisable = false
```
Comment (MaxPoon) [inline client/pages/me/me.js]: Missing semicolon. What is this variable for? How is it different from the one in `data`?

File: client/pages/me/me.js (hunk 14 above)
Comment (MaxPoon) [reply client/pages/me/me.js]: Should we still post the data even if btnDisable is set to true?

File: client/app.js [hunk 24]
```diff
@@ -11,7 +11,28 @@ App({
     qcloud.login({
       success(res) {
         that.userInfo = res;
+        qcloud.request({
+          url: config.service.userUrl + '?OpenId=' + that.userInfo.openId,
+          method: 'GET', 
+          success: function (res) {
+            //able to get user's personal information from server
+            that.userInfo = res.body;
```
Comment (MaxPoon) [inline client/app.js]: This will overwrite the original userInfo. You probably don't want to lose the openID, avatar url, etc... I think what we should do is to add the information from response into `that.userInfo`.

**Review by MaxPoon — CHANGES_REQUESTED** [review]

File: client/config.js (hunk 16 above)
Comment (ZhangYihanNus) [reply client/config.js]: So do we only change the local url for testing on our own computer?

Comment (MaxPoon) [reply client/config.js]: Yes

File: client/pages/me/me.js (hunk 20 above)
Comment (ERICCYS) [reply client/pages/me/me.js]: I think the method should be "PUT" instead.

**Review by ERICCYS — CHANGES_REQUESTED** [review]

File: client/pages/me/me.js (hunk 20 above)
Comment (ZhangYihanNus) [reply client/pages/me/me.js]: But isn't "PUT" used for only updating and not creating new items?

Comment (MaxPoon) [reply client/pages/me/me.js]: What we're doing at this endpoint is : create if not exists and update if it exists. It's an idempotent method. Using put should be fine.
https://stackoverflow.com/questions/630453/put-vs-post-in-rest
https://en.wikipedia.org/wiki/Idempotence

File: client/pages/me/me.wxml (hunk 22 above)
Comment (ZhangYihanNus) [reply client/pages/me/me.wxml]: If so we will not be able to make "Tap here to enter" text disappear when users tap to input (as far as I know). I think this can improve user experience but I'm okay with both ways.

File: client/pages/me/me.js (hunk 23 above)
Comment (ZhangYihanNus) [reply client/pages/me/me.js]: This is a mistake

File: client/app.js (hunk 24 above)
Comment (ZhangYihanNus) [reply client/app.js]: When "btnDisable" is set to true, the button will be disabled and function "infoSaveClick" won't run. When "btnDisable" is set to false, it will first store the info locally (for placeholders) and then POST. "btnDisable" will be set to true again in the first step.

Comment (MaxPoon) [reply client/app.js]: Got it!

File: client/pages/me/me.wxml (hunk 22 above)
Comment (MaxPoon) [reply client/pages/me/me.wxml]: You mean the placeholder will still be there even after the user inputs the text?

File: client/pages/me/me.js (hunk 20 above)
Comment (ZhangYihanNus) [reply client/pages/me/me.js]: Got it!

File: client/pages/me/me.wxml (hunk 22 above)
Comment (ZhangYihanNus) [reply client/pages/me/me.wxml]: If we use value instead of placeholder, users must first delete manually the text "Tap here to enter" as it is seen as the initial input value and is showing in the input box.

File: client/config.js (hunk 16 above)
Comment (MaxPoon) [reply client/config.js]: It's still the local url

File: client/pages/me/me.wxml (hunk 22 above)
Comment (MaxPoon) [reply client/pages/me/me.wxml]: You can set the value to the string it retrieved from the server, and still set the placeholder to "Tap here to enter". So if the value is empty, the placeholder will be shown.

File: client/pages/me/me.js [hunk 25]
```diff
@@ -1,66 +1,93 @@
...
-
-  },
-
-  /**
-   * 用户点击右上角分享
-   */
-  onShareAppMessage: function () {
-
+  infoSaveClick: function (event) {
+    this.setData({
```
Comment (MaxPoon) [inline client/pages/me/me.js]: The data should be set only after the put request is successful.

File: client/pages/me/me.js [hunk 26]
```diff
@@ -1,66 +1,93 @@
...
+    //update the latest info into server
+    qcloud.request({
+      url: 'http://localhost:5757',
+      //url: config.service.userUrl,
+      method: 'PUT',
+      // header: {
+      //   'content-type':'application/x-www-form-urlencoded'
+      // },
+      data: {
+        OpenID: this.data.personal.openId,
```
Comment (MaxPoon) [inline client/pages/me/me.js]: `OpenId`, not `OpenID`

File: client/pages/me/me.js [hunk 27]
```diff
@@ -1,66 +1,93 @@
...
+      chineseName: this.data.chineseNameInputValue,
+      englishName: this.data.englishNameInputValue,
+      phone: this.data.phoneInputValue,
+      email: this.data.emailInputValue,
+      btnDisable: true
+    });
+    //update the latest info into server
+    qcloud.request({
+      url: 'http://localhost:5757',
+      //url: config.service.userUrl,
```
Comment (MaxPoon) [inline client/pages/me/me.js]: This url should be used instead

File: client/config.js (hunk 16 above)
Comment (ZhangYihanNus) [reply client/config.js]: I was trying out the server on my PC, forgot to change it, will look out next time

File: client/pages/me/me.wxml (hunk 22 above)
Comment (ZhangYihanNus) [reply client/pages/me/me.wxml]: okay!

File: client/pages/me/me.js [hunk 28]
```diff
@@ -1,66 +1,85 @@
...
+      },
+      data: {
+        OpenId: this.data.personal.openId,
+        ChineseName: this.data.chineseNameInputValue,
+        EnglishName: this.data.englishNameInputValue,
+        MobileNumber: this.data.phoneInputValue,
+        EmailAddress: this.data.emailInputValue
+      },
+      success: function (res) {
+        //able update userinfo into server
```
Comment (MaxPoon) [inline client/pages/me/me.js:70]: update `app.userInfo` here

File: client/pages/me/me.wxml [hunk 29]
```diff
@@ -1,2 +1,50 @@
...
+    <image  class="userinfo-avatar" src="{{personal.avatarUrl}}" mode="cover"></image>
+    <text class="userinfo-nickname">{{personal.nickName}}</text>
+  </view>
+
+
+  <view class="body">   <!-- info -->
+    <view class="info">   <!-- Chinese name -->
+      <text>Chinese name</text>
+      <view class="infoInput">
+        <input wx:if="{{!personal.data.ChineseName}}" placeholder='Tap here to enter' bindinput='chineseNameInput'/>
```
Comment (MaxPoon) [inline client/pages/me/me.wxml:15]: I think just `<input placeholder='Tap here to enter' value='{{personal.data.ChineseName}}' 
 bindinput='chineseNameInput'/>` should work, don't need two input fields. Can you try it?

File: client/pages/me/me.js [hunk 30]
```diff
@@ -1,66 +1,85 @@
...
+    }
   },
-
-  /**
-   * 生命周期函数--监听页面显示
-   */
-  onShow: function () {
-
+  englishNameInput: function (event) {
+    if (event.detail.value.length > 0) {
```
Comment (MaxPoon) [inline client/pages/me/me.js:29]: If the original value is `Max`, and I'm deleting the three characters one by one before I click save, the final value of `englishNameInputValue` will be 'M', but the input field is empty on the screen. This isn't what we want.
Instead of not changing the value, I think we can disable the button if length is 0:
```
this.setData({
  englishNameInputValue: event.detail.value,
  btnDisable: event.detail.value.length === 0
});
```

**Review by MaxPoon — APPROVED** [review]
Overall it LGTM. But please see my suggestions before merging the PR.

**Review by ERICCYS — APPROVED** [review]
